Update of graphql definition for using ID fields + character connection.

Vassyli opened the pull request:
Also updated graphql packages to be up-to-date.

Vassyli committed:
Changes for the client.

Vassyli committed:
Changed character field on user to be a connection.

Vassyli committed:
Merge fixes

Vassyli committed:
Updated documentation and cleaned up code

austenmc commented:
Is this BaseConnection stuff not built in? This feels like boilerplate Relay stuff.

austenmc commented on src/AppBundle/GraphQL/Mutation/CharacterMutation.php:
```diff
@@ -36,7 +39,8 @@ function createCharacter(string $userId, string $characterName)
         return [
-            "character" => new CharacterType($this->getGame(), $character)
+            "characterEdge" => CharacterConnection::createEdgeFor($user, new CharacterType($this->getGame(), $character)),
+            "user" => new UserType($this->getGame(), $user),
```
Why return user?

Vassyli replied:
I have not tried to not return the user - I've been looking around for RANGE_ADD examples and they all return the parent field (user) of the connection + the connection edge in the mutation payload.

Vassyli replied:
Relay asks specifically for this field during the mutation which it probably wouldn't if it isn't required.

austenmc commented on src/AppBundle/GraphQL/Resolver/CharacterResolver.php:
```diff
@@ -18,13 +20,42 @@ class CharacterResolver extends BaseManagerService implements ContainerAwareInte
+        $cursor = $args["cursor"];
+        $user = $args["__data"];
+
+        $offset = CharacterConnection::decodeCursor($cursor);
```
Is there error handling needed? If so, let's at least file an issue or put a TODO so we remember to come back.

Vassyli replied:
Yes there is, thank you for pointing it out. I will add tests now anyway for both invalid and valid stuff. Incorrect cursor should throw an exception.

Vassyli commented:
Honestly, I have no clue. There is the Pagination helper, but I couldn't figure out how to use it, thought it's faster to implement it myself. I like this BaseConnection a lot more since it helps to understand the connection between the graphql definition and the class itself.

Vassyli committed:
Fixed a few tests and the test script to properly return the exit code.

Vassyli committed:
Fix graphiql controller wrapper

Vassyli committed:
Change deprecated Mutation builder to Relay::Mutation

Changed test script to now properly display colours, too.

Vassyli committed:
Adds missing tests for graphql types + introduces a few fixes

- Adds test for ActionGroupType
- Adds test for ActionType
- Improvements over CharacterType: Removed static method fromId and fromName and replaced them with the appropriate service calls. Tests were added to test the CharacterType class.
- Adds test for ConfigurationType
- Adds test for LibraryType and adds error handling for module author list if details are missing.
- Adds tests for RealmType
- Adds test for SessionType
- Added tests for UserType and removed static UserType methods. UserType::fromId and UserType::fromName were already covered by the UserManager service. Resolver's have access to services.
- Adds tests for ViewpointType

General rule: types should never have static method to create itself from parameters!

Vassyli committed:
Modified t to test everything, but accepting additional command line arguments

Vassyli committed:
Adds composer.lock file to lock versions of packages.

Vassyli committed:
Fixed duplicate db_name key issue for config.yml

Vassyli committed:
Adds tests for Resolvers

- Adds test for CharacterResolver
- Adds tests for RealmResolver
- Adds tests for SessionResolver
- Adds tests for UserResolver
- Adds tests for ViewpointResolver

Additionally, this commit fixed a few bugs and introduces error handling for invalid cursors.

Vassyli committed:
Adds test for BaseConnection

+ Adds support for backwards scrolling (last and before parameters).
